src/monte_carlo_tree_search.h

```diff
@@ -1,47 +1,56 @@
 #ifndef MCTS_H
 #define MCTS_H
 
-#include <torch/torch.h>
 #include <game.h>
 #include <model.h>
+#include <torch/torch.h>
+
 #include <random>
 
 class Node {
-public:
-    Node(float prior, int toPlay, int action);
-    bool IsExpanded();
-    float GetValue();
-    int GetPlayerId() { return _toPlay; };
-    int GetVisitCount() { return _visitCount; };
-    int SelectAction(float temperature);
-    Node* SelectChild();
-    void Expand(std::vector<int> state, int toPlay, std::vector<float> actionProbs);
-    std::vector<int> GetState() { return _state; };
-    int GetAction() { return _action; };
-    void AccumulateValue(float val) { _valueSum += val; };
-    void IncrementVisitCount() { _visitCount++; };
-private:
-    int _visitCount = 0;
-    int _toPlay = 0;
-    int _action = -1;
-    float _prior = 0;
-    float _valueSum = 0;
-    std::vector<Node*> _children;
-    std::vector<int> _state;
-    std::default_random_engine _generator;
-    float _UcbScore(Node parent, Node child);
+ public:
+  Node(float prior, int toPlay, int action);
+
+  int GetVisitCount() const { return visit_count_; };
+  int GetPlayerId() const { return to_play_; };
+  int GetAction() const { return action_; };
+  std::vector<int> GetState() const { return state_; };
+  void AccumulateValue(float val) { value_sum_ += val; };
+  void IncrementVisitCount() { ++visit_count_; };
+
+  void Expand(const std::vector<int>& state, int to_play,
+              const std::vector<float>& action_probs);
+  bool IsExpanded();
+  float GetValue();
+  int SelectAction(float temperature);
+  Node* SelectChild();
+
+ private:
+  int visit_count_ = 0;
+  int to_play_ = 0;
+  int action_ = -1;
+  float prior_ = 0;
+  float value_sum_ = 0;
+  std::vector<Node*> children_;
```

Comment:
Can you try making this `unique_ptr<Node>`? It might require you to implement copy constructors, move constructors etc.

If you don't feel like doing all that work that's fine too I can look at it later.